admin/config-ui/class-configuration-storage.php

```diff
@@ -0,0 +1,170 @@
+<?php
+/**
+ * @package WPSEO\Admin\ConfigurationUI
+ */
+
+/**
+ * Class WPSEO_Configuration_Storage
+ */
+class WPSEO_Configuration_Storage {
+
+	/** @var WPSEO_Configuration_Options_Adapter */
+	protected $adapter;
+
+	/** @var array WPSEO_Config_Field */
+	protected $fields = array();
+
+	/**
+	 * Add default fields
+	 */
+	public function add_default_fields() {
+		$this->add_field( new WPSEO_Config_Field_Upsell_Configuration_Service() );
+		$this->add_field( new WPSEO_Config_Field_Upsell_Site_Review() );
+		$this->add_field( new WPSEO_Config_Field_Success_Message() );
+		$this->add_field( new WPSEO_Config_Field_Mailchimp_Signup() );
+		$this->add_field( new WPSEO_Config_Field_Environment() );
+		$this->add_field( new WPSEO_Config_Field_Site_Type() );
+		$this->add_field( new WPSEO_Config_Field_Tag_Line() );
+		$this->add_field( new WPSEO_Config_Field_Multiple_Authors() );
+		$this->add_field( new WPSEO_Config_Field_Site_Name() );
+		$this->add_field( new WPSEO_Config_Field_Separator() );
+
+		$this->add_field( new WPSEO_Config_Field_Profile_URL_Facebook() );
+		$this->add_field( new WPSEO_Config_Field_Profile_URL_Twitter() );
+		$this->add_field( new WPSEO_Config_Field_Profile_URL_Instagram() );
+		$this->add_field( new WPSEO_Config_Field_Profile_URL_LinkedIn() );
+		$this->add_field( new WPSEO_Config_Field_Profile_URL_MySpace() );
+		$this->add_field( new WPSEO_Config_Field_Profile_URL_Pinterest() );
+		$this->add_field( new WPSEO_Config_Field_Profile_URL_YouTube() );
+		$this->add_field( new WPSEO_Config_Field_Profile_URL_GooglePlus() );
+	}
+
+	/**
+	 * Allow for field injections
+	 *
+	 * @param WPSEO_Config_Field $field Field to add to the stack.
+	 */
+	public function add_field( WPSEO_Config_Field $field ) {
+		$this->fields[] = $field;
+	}
+
+	/**
+	 * Set the adapter to use
+	 *
+	 * @param WPSEO_Configuration_Options_Adapter $adapter Adapter to use.
+	 */
+	public function set_adapter( WPSEO_Configuration_Options_Adapter $adapter ) {
+		$this->adapter = $adapter;
+	}
+
+	/**
+	 * Retrieve the current adapter
+	 *
+	 * @return WPSEO_Configuration_Options_Adapter
+	 */
+	public function get_adapter() {
+		return $this->adapter;
+	}
+
+	/**
+	 * Retrieve the registered fields
+	 *
+	 * @returns array List of settings.
+	 */
+	public function retrieve() {
+		$output = array();
+
+		/** @var WPSEO_Config_Field $field */
+		foreach ( $this->fields as $field ) {
+			$build = array(
```

Comment:
My first thought here was, shouldn't the `WPSEO_Config_Field` class not be responsible for knowing it's JSON representation? My second thought is: Why not make it implement the [JsonSerializable](http://php.net/manual/en/class.jsonserializable.php) interface and in this function just do `return json_encode( $this->fields );`?